contracts/AINFT721.sol

```diff
@@ -0,0 +1,269 @@
+
+    constructor() ERC721("Name", "SYMBOL") {
+        _grantRole(DEFAULT_ADMIN_ROLE, msg.sender);
+        _grantRole(PAUSER_ROLE, msg.sender);
+        _grantRole(MINTER_ROLE, msg.sender);
+    }
+
+    function supportsInterface(
+        bytes4 interfaceId
+    )
+        public
+        view
+        override(
+            AccessControl,
+            ERC721Enumerable,
+            ERC721,
+            IERC165
+        )
+        returns (bool)
+    {
+        return super.supportsInterface(interfaceId);
+    }
+
+    function pause() public onlyRole(PAUSER_ROLE) {
+        _pause();
+    }
+
+    function unpause() public onlyRole(PAUSER_ROLE) {
+        _unpause();
+    }
+
+    function safeMint(
+        address to
+    )
+        public
+        onlyRole(MINTER_ROLE)
+    {
+        uint256 tokenId = _tokenIdCounter.current();
+        _tokenIdCounter.increment();
+        _safeMint(to, tokenId);
+    }
+
+    function _beforeTokenTransfer(
+        address from,
+        address to,
+        uint256 tokenId,
+        uint256 batchSize
+    )
+        internal
+        override(ERC721, ERC721Enumerable)
+        whenNotPaused
+    {
+        super._beforeTokenTransfer(from, to, tokenId, batchSize);
+    }
+
+    // The following functions are overrides required by Solidity.
+    function _burn(uint256 tokenId) internal override(ERC721) {
+        super._burn(tokenId);
+    }
+
+    ////
+    // URI VIEW FUNCTIONS
+    ////
+
+    function _baseURI() internal view override returns (string memory) {
+        return baseURI;
+    }
+
+    /**
+     * @dev Reverts if the `tokenId` has not been minted yet.
+     */
+    function _requireMinted(uint256 tokenId) internal view virtual override {
+        super._requireMinted(tokenId);
+    }
+
+    function tokenURI(
+        uint256 tokenId
+    ) public view override(ERC721) returns (string memory) {
+        _requireMinted(tokenId);
+        return getRecentTokenURI(tokenId);
+    }
+
+    function _getMetadataStorageKey(
+        uint256 tokenId,
+        uint256 version
+    ) internal pure returns (bytes32) {
+        return
+            keccak256(
+                bytes(
+                    string(
+                        abi.encodePacked(
+                            tokenId.toString(),
+                            "AINFT delimeter",
+                            version.toString()
+                        )
+                    )
+                )
+            );
+    }
+
+    function getMetadataStorage(
+        uint256 tokenId
+    ) public view returns (MetadataContainer memory) {
+        //TODO
+        uint256 currentVersion = _tokenURICurrentVersion[tokenId];
+        bytes32 key = _getMetadataStorageKey(tokenId, currentVersion);
+
+        return _metadataStorage[key];
+    }
+
+    function getTokenURICurrentVersion(
+        uint256 tokenId
+    ) public view returns (uint256) {
+        return _tokenURICurrentVersion[tokenId];
```

Comment:
`_requireMinted(tokenId)`?

Reply:
Oh I missed it. I'll put the pre-check function